x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/ClaimSetting.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.support;
+
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.RealmSettings;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+/**
+ * The OIDC and JWT realms offer a number of settings that rely on claim values that are populated by the JWT issuer. In the OIDC realm,
+ * the issuer is the OIDC OP, and claims can be added from the User Info response too.
+ * Each claim has 2 settings:
+ * <ul>
+ * <li>The name of the JWT claim to use</li>
+ * <li>An optional java pattern (regex) to apply to that claim value in order to extract the substring that should be used.</li>
+ * </ul>
+ * For example, the Elasticsearch User Principal could be configured to come from the OpenID Connect standard claim "email",
+ * and extract only the local-part of the user's email address (i.e. the name before the '@').
+ * This class encapsulates those 2 settings.
+ */
+public final class ClaimSetting {
+    public static final String CLAIMS_PREFIX = "claims.";
+    public static final String CLAIM_PATTERNS_PREFIX = "claim_patterns.";
+
+    private final Setting.AffixSetting<String> claim;
+    private final Setting.AffixSetting<String> pattern;
+
+    // type should be one of OpenIdConnectRealmSettings.TYPE="oidc" or JwtRealmSettings.TYPE="jwt"
```

Comment:
Should we throw if it isn't then ?

Reply:
I think we should add null and empty string checks, but no "oidc" or "jwt" checks.

If we limit reuse of this class to only "oidc" or "jwt" only, Elasticsearch can add extra types to remove it for more realms in future. If a customer tries to reuse this class in their own custom realms, the "oidc" or "jwt" check will be to restrictive.

Reply:
I would like to discuss what to add. I ran into issues testing for null and empty string.
- If no default specified, null seems to be interpreted as empty string.
- If default specified, null and empty are ignored, replaced by default value.

Reply:
Added null and empty checks. Added regex check. Discussed corner issues. Resolving conversation.